### PR TITLE
Feature/sites appear in launcher with icons

### PR DESCRIPTION
**Pull request by ofekengel**
added three main abilities:
  - links will now appear in jupyterlab's launcher
  - it will try to retrieve the site's icon through its DOM (sometimes fails due to it being in a js file or something)
  - added a (more) accurate way of determining whether a site can be a direct source in an iframe or not

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,6 +28,8 @@ install the server extension, and add the following to `jupyter_notebook_config.
 ```python3
 c.JupyterLabIFrame.iframes = ['list', 'of', 'sites']
 ```
+The sites will appear in the launcher as the site's icon. Some icons might not be retrieved successfully
+![](https://raw.githubusercontent.com/ofekengel/jupyterlab_iframe/feature/sites-appear-in-launcher-with-icons/docs/static_links_on_launcher.PNG)
```
Comment (timkpaine) [inline README.md:32]: lower case `.png`

File: js/package.json [hunk 2]
```diff
@@ -40,11 +40,18 @@
     "@jupyterlab/application": "^3.0.0",
     "@jupyterlab/apputils": "^3.0.0",
     "@jupyterlab/coreutils": "^5.0.0",
+    "@jupyterlab/launcher": "^3.0.2",
     "@lumino/disposable": "^1.1.2",
-    "requests-helper": "^0.1.5"
+    "@types/node": "^14.14.22",
+    "axios": "^0.21.1",
```
Comment (timkpaine) [inline js/package.json:46]: remove dep

File: js/package.json [hunk 3]
```diff
@@ -40,11 +40,18 @@
...
+    "image-pixels": "^2.2.2",
```
Comment (timkpaine) [inline js/package.json:47]: remove dep

File: js/package.json [hunk 4]
```diff
@@ -40,11 +40,18 @@
...
+    "imagetracerjs": "^1.2.6",
```
Comment (timkpaine) [inline js/package.json:48]: remove dep

File: js/src/SVGAssets.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
...410156-85.703125-46.402344-108.28125-107.515625zm104.488281 8.105469c29.660156-.390625 59.242188-3.101562 88.480469-8.105469l19.800781-3.425781c-22.578125 61.128906-62.417969 103.136719-108.28125 107.523438zm85.753906-23.875c-28.335937 4.847656-57.007812 7.464844-85.753906 7.832031v-103.550781h127.878906c-.535156 30.769531-5.203125 61.324219-13.878906 90.847656zm58.117188-111.71875c-.527344-29.691406-4.746094-59.203125-12.558594-87.855469l68.253906 11.777344c9.835938 24.207031 15.332032 49.960937 16.234375 76.078125zm47.601562-93.710938-65.425781-11.289062c-11.761719-38.371094-33.765625-72.808594-63.648437-99.601562 55.878906 18.648437 102.21875 58.457031 129.074218 110.890624zm-269.871094-110.890624c-29.882812 26.792968-51.886718 61.230468-63.648437 99.601562l-65.425781 11.289062c26.855468-52.433593 73.195312-92.242187 129.074218-110.890624zm-129.074218 314.3125 65.425781 11.289062c11.761719 38.371094 33.765625 72.808594 63.648437 99.601562-55.878906-18.648437-102.21875-58.457031-129.074218-110.890624zm269.871094 110.890624c29.882812-26.792968 51.886718-61.230468 63.648437-99.601562l65.425781-11.289062c-26.855468 52.433593-73.195312 92.242187-129.074218 110.890624zm0 0\"/></svg>";
```
Comment (timkpaine) [inline js/src/SVGAssets.ts:2]: move to svg file

File: js/src/SVGAssets.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+const openSiteSVG = "<?xml version=\"1.0\" encoding=\"iso-8859-1\"?>\n" +
```
Comment (timkpaine) [inline js/src/SVGAssets.ts:3]: move to svg file

File: js/src/SVGAssets.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    "<!-- Generator: Adobe Illustrator 19.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->\n" +
+    "<svg version=\"1.1\" id=\"Capa_1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\"\n" +
+    "\t viewBox=\"0 0 294.834 294.834\" style=\"enable-background:new 0 0 294.834 294.834;\" xml:space=\"preserve\">\n" +
+    "<g>\n" +
+    "\t<path d=\"M288.834,0H47.413c-3.313,0-6,2.687-6,6v17.398c0,3.313,2.687,6,6,6s6-2.687,6-6V12h229.421v229.421h-11.398\n" +
+    "\t\tc-3.313,0-6,2.687-6,6s2.687,6,6,6h17.398c3.313,0,6-2.687,6-6V6C294.834,2.687,292.147,0,288.834,0z\"/>\n" +
+    "\t<path d=\"M247.421,41.413H6c-3.313,0-6,2.687-6,6v241.421c0,3.313,2.687,6,6,6h241.421c3.313,0,6-2.687,6-6V47.413\n" +
+    "\t\tC253.421,44.1,250.734,41.413,247.421,41.413z M241.421,282.834H12V53.413h229.421V282.834z\"/>\n" +
+    "</g>\n" +
+    "<g>\n" +
```
Comment (timkpaine) [inline js/src/SVGAssets.ts:13]: move extras

File: js/src/faviconFetcher.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const imagedataToSVG = require("imagetracerjs").imagedataToSVG;
```
Comment (timkpaine) [inline js/src/faviconFetcher.ts:2]: switch to `import`

File: js/src/faviconFetcher.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const pixels = require("image-pixels");
```
Comment (timkpaine) [inline js/src/faviconFetcher.ts:4]: switch to `import`

File: js/src/faviconFetcher.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const scale = require("scale-that-svg");
```
Comment (timkpaine) [inline js/src/faviconFetcher.ts:6]: switch to `import`

File: js/src/faviconFetcher.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+const getSiteDom = async (url: string): Promise<Document> => fetch(url).then(async (res) => {
+  const parser = new DOMParser();
+  return parser.parseFromString(await res.text(), "text/xml");
+});
+
+const getFavicon = async (site: string): Promise<string> => {
+  let faviconUri: string;
+  const dom = await getSiteDom(site);
+  const nodeList = dom.getElementsByTagName("link");
+  // eslint-disable-next-line @typescript-eslint/prefer-for-of
```
Comment (timkpaine) [inline js/src/faviconFetcher.ts:25]: listen to eslint

File: js/src/faviconFetcher.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    throw new FaviconNotFoundError(site);
+  }
+  let siteUrl = new URL(site);
+  if (siteUrl.pathname === "/iframes/proxy"){
+    const faviconUrl = `${new URL(faviconUri, siteUrl.searchParams.get("path")).href}`;
+    siteUrl.searchParams.set("path", faviconUrl);
+  } else {
+    siteUrl = new URL(faviconUri);
+  }
+  const data = await pixels(siteUrl.href);
```
Comment (timkpaine) [inline js/src/faviconFetcher.ts:42]: remove dep

File: js/src/faviconFetcher.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  const unscaledSvg = imagedataToSVG(data);
```
Comment (timkpaine) [inline js/src/faviconFetcher.ts:43]: remove dep

File: js/src/iframeUtils.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import axios, {AxiosResponse} from "axios";
```
Comment (timkpaine) [inline js/src/iframeUtils.ts:1]: remove dep

File: js/src/index.ts [hunk 15]
```diff
@@ -202,12 +213,15 @@ function activate(app: JupyterFrontEnd, palette: ICommandPalette) {
...
       }
     },
+    icon: openSiteIcon,
     isEnabled: () => true,
     label: "Open IFrame",
   });
 
   // Add the command to the palette.
   palette.addItem({command: open_command, category: "Sites"});
+  launcher.add({command: open_command, rank: 1, category: "Sites"});
```
Comment (timkpaine) [inline js/src/index.ts:223]: configure launcher addition from config (can follow how its done for `welcome`)

File: js/src/index.ts [hunk 16]
```diff
@@ -228,7 +242,7 @@ function activate(app: JupyterFrontEnd, palette: ICommandPalette) {
           welcome_included = true;
         }
         if (site) {
-          registerSite(app, palette, site);
+          registerSite(app, palette, site, launcher);
```
Comment (timkpaine) [inline js/src/index.ts:245]: configure launcher addition from config (can follow how its done for `welcome`)

File: js/src/index.ts [hunk 17]
```diff
@@ -242,14 +256,14 @@ function activate(app: JupyterFrontEnd, palette: ICommandPalette) {
           welcome_included = true;
         }
         if (actual_site) {
-          registerSite(app, palette, actual_site);
+          registerSite(app, palette, actual_site, launcher);
```
Comment (timkpaine) [inline js/src/index.ts:259]: configure launcher addition from config (can follow how its done for `welcome`)

File: js/src/index.ts [hunk 18]
```diff
@@ -242,14 +256,14 @@ function activate(app: JupyterFrontEnd, palette: ICommandPalette) {
...
         }
       }
 
 
       if (!welcome_included) {
         if (welcome !== "") {
-          registerSite(app, palette, welcome);
+          registerSite(app, palette, welcome, launcher);
```
Comment (timkpaine) [inline js/src/index.ts:266]: configure launcher addition from config (can follow how its done for `welcome`)

File: README.md [hunk 19]
```diff
@@ -28,6 +28,8 @@ install the server extension, and add the following to `jupyter_notebook_config.
...
```
Comment (timkpaine) [inline README.md:33]: demo how to disable adding to launcher via config

**Review by timkpaine — CHANGES_REQUESTED** [review]

**Comment by timkpaine** [comment]
I will start up a branch to start on the changes

**Comment by timkpaine** [comment]
https://github.com/timkpaine/jupyterlab_iframe/pull/75

File: js/src/SVGAssets.ts (hunk 5 above)
Comment (ofekengel) [reply js/src/SVGAssets.ts:2]: note that in order to display an icon in launcher you need to supply jupyterlab with a string of an SVG ([source](https://discourse.jupyter.org/t/use-custom-icon-on-launcher-item/5403/2))

File: js/package.json (hunk 3 above)
Comment (ofekengel) [reply js/package.json:47]: needed to convert png/jpg icons to svg's jupyterlab can display in launcher

File: js/package.json (hunk 4 above)
Comment (ofekengel) [reply js/package.json:48]: needed to convert png/jpg icons to svg's jupyterlab can display in launcher

File: js/package.json (hunk 3 above)
Comment (timkpaine) [reply js/package.json:47]: nope, you can use `kernelIconUrl`: https://github.com/jupyterlab/jupyterlab/blob/60336e5c8d600f609771a78fbeedb50c5dc070e5/packages/launcher/src/index.tsx#L357

I use a `png` in this library (as an example): https://github.com/jpmorganchase/jupyterlab_templates/blob/main/js/src/index.ts#L175

File: js/package.json (hunk 4 above)
Comment (timkpaine) [reply js/package.json:48]: see above

File: js/src/SVGAssets.ts (hunk 5 above)
Comment (timkpaine) [reply js/src/SVGAssets.ts:2]: see above

File: js/package.json (hunk 3 above)
Comment (ofekengel) [reply js/package.json:47]: Cool, that would have saved me some time 🙃

Comment (timkpaine) [reply js/package.json:47]: @ofekengel yeah its not documented anywhere I don't think...it took a lot of digging to figure out how to make a "fake" kernel launcher :joy:

File: js/src/SVGAssets.ts (hunk 5 above)
Comment (timkpaine) [reply js/src/SVGAssets.ts:2]: I've responded to that thread https://discourse.jupyter.org/t/use-custom-icon-on-launcher-item/5403/2

Comment (ofekengel) [reply js/src/SVGAssets.ts:2]: amazing 😄

**Pull request closed without merging**